File: elixir/lib/2021/10_syntax_scoring/aoc202110.ex
defmodule AOC2021.Day10 do
  @moduledoc """
  Advent of Code 2021, day 10: Syntax Scoring
  """
  require AOC

  @pairs %{?( => ?), ?[ => ?], ?{ => ?}, ?< => ?>}
  @scores %{?) => 3, ?] => 57, ?} => 1197, ?> => 25_137}
  @chunk_scores %{?) => 1, ?] => 2, ?} => 3, ?> => 4}

  @doc """
  Parse input
  """
  def parse(puzzle_input),
    do: puzzle_input |> String.split("\n") |> Enum.map(&String.to_charlist/1)

  @doc """
  Solve part 1
  """
  def part1(input), do: input |> Enum.map(&illegal_char/1) |> score_chars() |> Enum.sum()

  @doc """
  Solve part 2
  """
  def part2(input) do
    input
    |> Enum.map(&illegal_char/1)
    |> Enum.reject(&Map.has_key?(@scores, &1))
    |> Enum.map(&score_chunk/1)
    |> Statistics.median()
  end

  @doc """
  Find the first illegal character in a corrupt chunk

  ## Example:

      iex> illegal_char('({}>)')
      ?>
  """
  def illegal_char(chunk), do: illegal_char(chunk, [])
  def illegal_char([], seen), do: seen |> Enum.reverse() |> close_chunk()

  def illegal_char([char | chunk], []),
    do: if(Map.has_key?(@pairs, char), do: illegal_char(chunk, [char]), else: char)

  def illegal_char([char | chunk], [prev | seen]) do
    cond do
      Map.has_key?(@pairs, char) -> illegal_char(chunk, [char, prev | seen])
      char == @pairs[prev] -> illegal_char(chunk, seen)
      true -> char
    end
  end

  @doc """
  Close the given chunk

  ## Example:

      iex> close_chunk('<{([')
      '])}>'
  """
  def close_chunk(chunk), do: chunk |> Enum.reduce([], fn char, acc -> [@pairs[char] | acc] end)

  @doc """
  Score illegal characters

  ## Example:

      iex> score_chars([?), ?], ')}])', ?}, ?>, ?)])
      [3, 57, 0, 1197, 25137, 3]
  """
  def score_chars(characters), do: characters |> Enum.map(&Map.get(@scores, &1, 0))

  @doc """
  Score closing chunks

  ## Example:

      iex> score_chunk('])}>')
      294
  """
  def score_chunk(chunk),
    do: chunk |> Enum.reduce(0, fn char, sum -> 5 * sum + @chunk_scores[char] end)

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2021/10_syntax_scoring/aoc202110.py
"""AoC 10, 2021: Syntax Scoring."""

# Standard library imports
import pathlib
import statistics
import sys

PAIRS = {"(": ")", "[": "]", "{": "}", "<": ">"}


def parse_data(puzzle_input):
    """Parse input."""
    return puzzle_input.split("\n")


def part1(data):
    """Solve part 1."""
    return sum(score_char(illegal_char(chunk)) for chunk in data)


def part2(data):
    """Solve part 2."""
    chunks = [
        left_over
        for chunk in data
        if (left_over := illegal_char(chunk)) not in PAIRS.values()
    ]
    return statistics.median([score_chunk(chunk) for chunk in chunks])


def illegal_char(chunk):
    """Find the first illegal character, or close up chunk if there are no
    illegal characters.

    >>> illegal_char("({}<)>")
    ')'

    >>> illegal_char("<([]<")
    ['>', ')', '>']
    """
    closing = []
    for char in chunk:
        match (char, closing):
            case char, closing if char in PAIRS:
                closing.append(PAIRS[char])
            case char, [*_, prev] if char == prev:
                closing.pop()
            case char, _:  # Illegal character
                return char

    # No illegal characters, return closing chunk
    return closing[::-1]


def score_char(char):
    """Score illegal character.

    >>> score_char("}")
    1197

    >>> score_char(["]", ")", "}", ">"])
    0
    """
    match char:
        case ")":
            return 3
        case "]":
            return 57
        case "}":
            return 1197
        case ">":
            return 25137

    return 0


def score_chunk(chunk, score=0):
    """Score left over chunk.

    >>> score_chunk(["]", ")", "}", ">"])
    294
    """
    match chunk:
        case [")", *tail]:
            return score_chunk(tail, score * 5 + 1)
        case ["]", *tail]:
            return score_chunk(tail, score * 5 + 2)
        case ["}", *tail]:
            return score_chunk(tail, score * 5 + 3)
        case [">", *tail]:
            return score_chunk(tail, score * 5 + 4)
        case []:
            return score


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
